### PR TITLE
I2C: introduce and adapt new I2C interface (2nd attempt)

### PR DESCRIPTION
**Pull request by dylad**
### Contribution description

This PR introduces and adapts to all cpu/board/device drivers, the new I2C interface of RIOT from the feature branch. The branch used is a cloned and rebased version of new_i2c_if because I didn't want to rebase new_i2c_if yet.
All comments/reviews will be much appreciate as the PR is pretty big.

For the 2nd attempt, the new_i2c_if3 used branch is based on new_i2c_if which has been rebase without --preserve-merges option. This option seems to cause some trouble (see discussion #9539)

### Issues/PRs references
Tracking issue of the I2C refactoring #6577
Discussion #9539

### CODE REVIEW

**Comment by dylad** [comment]
@cladmi I have a huge problem here 
[this](https://github.com/RIOT-OS/RIOT/pull/9548/commits/f99c8aeff19243579bf8ac6a835aabd74e84261a) commit introduces i2C support in the Makefile.features for nucleo-f070 in the new_i2c_if branch.
But when I rebase the branch new_i2c_if3 to master I got [this](https://github.com/RIOT-OS/RIOT/commit/c6d345ce67410e1a50e2e246cbc835699318c129)
the added line to Makefile.features is now to nucleo-f030 not nucleo-f070.
I guess rebase didn't do its job correctly but nucleo's board were renamed.
What can I do ?

**Comment by cladmi** [comment]
My idea would be to try rebasing but forbid git to perform automatic merge, so that you still must verify each commit.
Maybe you can by putting "edit" for all commits except merges during the rebase and checking each commit… Not sure how "--preserve" likes this though.

**Comment by cladmi** [comment]
Something that can help finding these, is https://stackoverflow.com/a/40183550/395687

    git commit-tree riot/new_i2c_if3^{tree} -p riot/new_i2c_if  -p riot/master -m 'Is that ok?'

And doing `git show GIVEN_HASH` to see the difference between "merged" diff and "rebased" diff. Some lines have a `++` also. It can help narrowing down these issues.

**Comment by dylad** [comment]
@cladmi The I2C refactoring is now over and ready to be merged. The decision to add it to the current release belongs to you :)
Murdock is finally happy with this PR !

**Comment by cladmi** [comment]
Do you have somewhere the list of tests that was run so they could be re-run in master ?
At least the easy ones and the one that got conflicts.
I think it would be great to add feedback that the rebase is indeed successful.
Also if you could summarize the changes so I could add an info to the release.

**Comment by cladmi** [comment]
Question, what are all these "fix attr order" and "cleanup AF" ? are these meant to be squashed after ?

I may do a commit by commit diff review on monday.

**Comment by dylad** [comment]
@cladmi mentionned issues are CPP fixes because Murdock was complaining.
Regarding the tests, well all tests were run on hardware with sensors with the related PR. So It's not something easy ton re-test.

**Comment by kaspar030** [comment]
Is there somewhere  a summary of the actual API changes?

**Comment by dylad** [comment]
@kaspar030 You can have a look at #6577 
Or you can search with github all PRs labelled I2C REWORK, then you will have the complete list of all merged PRs into the feature branch.

**Comment by kaspar030** [comment]
> You can have a look at #6577
> Or you can search with github all PRs labelled I2C REWORK, then you will have the complete list of all merged PRs into the feature branch.

@dylad I'm looking for an API change description. There was a PR / issue somewhere where the discussion took place, but I think it is not #6577.

**Comment by kaspar030** [comment]
> I can open a wiki page for summarize all the changes adopted.

That would be awesome!

**Comment by dylad** [comment]
@kaspar030 I'll inform you when this wiki page will be written. I'll start this weekend if I find some time.

**Comment by dylad** [comment]
@kaspar030 please see [here](https://github.com/RIOT-OS/RIOT/wiki/I2C-rework) this is the first round of the wiki page. I am not a good writer usually so don't hesitate to correct me or ask for more information.

**Comment by aabadie** [comment]
not blocking this PR, but I just want to say that I'm against merging this PR in the current state, because of the suboptimal changes applied to the STM32F1 boards.

**Comment by aabadie** [comment]
@dylad, for the wiki page, this PR also contains other changes, not directly related to the new API. These are the ones that come to my mind right now:
- all STMs CPUs provide I2C and the driver is unified (well, there's 2 versions of it)
- STM32F3 driver is fixed (it's broken in master)
- all STMs based boards use the new config style (array of structs), instead of defines
- previous point is also valid for sam0
- an I2C config was added to some nucleo/st disco boards provide (b-l475e-iot01a, nucleo-f722ze, nucleo-f070rb, and maybe others I forgot)
- nrf52xxdk I2C pin change to match Arduino pinout
- API change in some device drivers: lsm303dlhc, etc (I don't remember the complete list and it's difficult for me to get it  on my phone)

**Comment by dylad** [comment]
I think #9574 should be handle before this one. Then I could rebase this branch on master.

**Comment by cladmi** [comment]
It is time so I will do the first release candidate without this PR.
We need to start testing.

I will still agree to backport it this week until the next RC, the branch was correctly tested before so it should be OK and it would save a lot of maintenance troubles.

However I would require that you re-run the automated tests (@MrKevinWeiss ?) and explain how some maintainers can re-test boards they have. Just have enough information to be confident.

@aabadie are your concerns addressed with the merge of #9574 and the AF handling of STM32f1 ?

**Comment by dylad** [comment]
Murdock is happy as expected. I can re-run tests on some boards if it's needed.

**Comment by kaspar030** [comment]
> I will still agree to backport it this week until the next RC, the branch was correctly tested before so it should be OK and it would save a lot of maintenance troubles.

It's your call as release manager, but would that be wise? With +31,755 −18,939 (and not changing vendor headers) this is probably the biggest PR ever.

**Comment by cladmi** [comment]
> >  I will still agree to backport it this week until the next RC, the branch was correctly tested before so it should be OK and it would save a lot of maintenance troubles.

> It's your call as release manager, but would that be wise? With +31,755 −18,939 (and not changing vendor headers) this is probably the biggest PR ever.

I am doing the RC today as tests should be started before some maintainers go on vacations, I could have done it in one week but would only have prevent starting to test.
It however means we might need to backport more and run tests on both master and the RC.

I agree it is a huge PR, but it has already been reviewed. Only not the rebase.
It is still not included, but I leave the possibility if tests are re-run.

Also I do not really feel comfortable about keeping the embargo over the release cycle. So I want to keep the possibility of merging it open.

**Comment by kaspar030** [comment]
> Also I do not really feel comfortable about keeping the embargo over the release cycle. So I want to keep the possibility of merging it open.

Actually, if the release is branched, this PR could be merged into master now and thus end the embargo.

**Comment by MrKevinWeiss** [comment]
Test results (format is board and commit hash)
[nucleo-f401re-347bcd225233f753cb01c798f3b7f407f865df48.txt](https://gist.github.com/MrKevinWeiss/edc8d0083800c165fed70467f1cac792)

[nucleo-l152re--347bcd225233f753cb01c798f3b7f407f865df48.txt]
(https://gist.github.com/MrKevinWeiss/5c60804fa5f70a94e9cebb526d59c98e)

[nucleo-l073rz-347bcd225233f753cb01c798f3b7f407f865df48.txt]
(https://gist.github.com/MrKevinWeiss/371696d0343a02f30d50f79dbabfb76a)
Seems like some problems when writing to address that is not there.  It recovers eventually though when writing a second missing address.

[nucleo-f103rb-347bcd225233f753cb01c798f3b7f407f865df48.txt](https://gist.github.com/MrKevinWeiss/fcefd544e25384b5bc947743c9d25216)
Hmmm seems like some big problems here.  I will have to investigate

[arduino-mega2560-347bcd225233f753cb01c798f3b7f407f865df48.txt](https://gist.github.com/MrKevinWeiss/93eecc9dd63eadb9621d3eaa6d42ce06)

Some boards I had some toolchain problems with (nrf52-dk, frdm-k22f)

To me it seems like the adaptation to the new API is good still some issues but they probably exist in master and I have the issue that collects all them.  My only concern is the nucleo-f103rb...

**Comment by dylad** [comment]
Run some tests on sam0 and nrf52dk with tests/drivers_xxx (ADXL345 / SI7021) looks like everything is fine.
@MrKevinWeiss Could you detailed your toolchain issue ? I don't have any problem with it.
I don't have ST boards for testing.

**Comment by MrKevinWeiss** [comment]
Oh no the problem is mine, I just never setup the toolchain/flasher for everything.  I would like to get this merged just so it stops people as the adaptation is good (as long as murdock works).  Here is my toolchain info is your just curious.

```
LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 16.04.4 LTS
Release:	16.04
Codename:	xenial
-----------------------------
             native gcc: gcc (Ubuntu 5.4.0-6ubuntu1~16.04.10) 5.4.0 20160609
      arm-none-eabi-gcc: arm-none-eabi-gcc (GNU Tools for Arm Embedded Processors 7-2017-q4-major) 7.2.1 20170904 (release) [ARM/embedded-7-branch revision 255204]
                avr-gcc: avr-gcc (GCC) 4.9.2
       mips-mti-elf-gcc: missing
             msp430-gcc: missing
   riscv-none-embed-gcc: missing
                  clang: clang version 5.0.0-3~16.04.1 (tags/RELEASE_500/final)

Installed compiler libs
-----------------------
   arm-none-eabi-newlib: "2.5.0"
    mips-mti-elf-newlib: missing
riscv-none-embed-newlib: missing
               avr-libc: "1.8.0svn" ("20111229")

Installed development tools
---------------------------
                  cmake: cmake version 3.5.1
               cppcheck: missing
                doxygen: missing
                 flake8: missing
                    git: git version 2.7.4
             coccinelle: missing
```

**Comment by MrKevinWeiss** [comment]
It appears the problem with the nucleo-f103rb exists in master.  When the I2C is enabled the sda and scl lines are held down (in both).  Since it is in master it doesn't effect the adaptation and all other boards tested seem to pass correctly.  I will give this my **ACK** to get merged.

@dylad @cladmi

File: boards/nucleo-f103rb/include/periph_conf.h [hunk 1]
```diff
@@ -147,35 +147,37 @@ static const uart_conf_t uart_config[] = {
...
-#define I2C_1_SCL_PIN       GPIO_PIN(PORT_B, 10)
-#define I2C_1_SDA_PIN       GPIO_PIN(PORT_B, 11)
+static const i2c_conf_t i2c_config[] = {
+    {
+        .dev            = I2C1,
+        .speed          = I2C_SPEED_NORMAL,
+        .scl_pin        = GPIO_PIN(PORT_B, 8),
+        .sda_pin        = GPIO_PIN(PORT_B, 9),
+        .scl_af         = GPIO_AF_OUT_OD,
+        .sda_af         = GPIO_AF_OUT_OD,
```
Comment (aabadie) [inline boards/nucleo-f103rb/include/periph_conf.h]: here and in many other places related to stm32f1, the af attributes should be removed. Is it possible to just rebase the branch and remove the related commits. I saw the right fix in another PR, the corresponding commit should be applied then.

Comment (dylad) [reply boards/nucleo-f103rb/include/periph_conf.h]: This has been fix on (new_i2c_if branch)[https://github.com/RIOT-OS/RIOT/blob/new_i2c_if/boards/nucleo-f103/include/periph_conf.h]
Looks like these values are re-introduced by the rebase because nucleo were renamed..
I'll try to find why again.

**Comment by dylad** [comment]
Sorry I forced push the branch once again, @aabadie's issue should be fix now.

**Comment by aabadie** [comment]
Now that this PR is rebased on top of master, I think 3ab6027 is not needed anymore (drop I2C feature from nucleo-f070rb)

**Comment by dylad** [comment]
@aabadie I'll redo some tests but this PR is needed because git rebase did the wrong choice during the rebase and introduce conflicts so I'd like to keep it for now.

File: boards/nucleo-f103rb/Makefile.features [hunk 2]
```diff
@@ -6,9 +6,6 @@ FEATURES_PROVIDED += periph_spi
 FEATURES_PROVIDED += periph_timer
 FEATURES_PROVIDED += periph_uart
 
-# load the common Makefile.features for Nucleo boards
```
Comment (aabadie) [inline boards/nucleo-f103rb/Makefile.features]: This change is unrelated. The commit that introduced it should be removed.

Comment (dylad) [reply boards/nucleo-f103rb/Makefile.features]: Done.

**Comment by MrKevinWeiss** [comment]
@dylad Anything I can do to help get this guy merged?

**Comment by dylad** [comment]
I guess we should give to @cladmi and others maintainers some tests results as proof ?

**Comment by cladmi** [comment]
It is also currently Release tests time so I do not have time to look at your comments updates or look at the PR.

**Comment by MrKevinWeiss** [comment]
Well for whenever people are ready I ran tests on the following boards and **THEY ALL PASS**:

- arduino-mega2560
- frdm-k22f
- nrf52-dk
- nucleo-f103rb
- nucleo-f401re
- nucleo-l152re
- samr21-xpro

The commit hash is 69297a09394d7d58ae868f472c5be0dccd971245

The results are [here](https://gist.github.com/MrKevinWeiss/d76a44a869f3ea67256c05f3d142ae54)

**Comment by smlng** [comment]
@kaspar030 care to approve or dismiss your review as the release branch is in place for some time now. We have lot's of stuff waiting for this to be master.

**Comment by kaspar030** [comment]
Alright, let's go.

**Comment by smlng** [comment]
ACK and GO then!

**Comment by kYc0o** [comment]
Woooow! Congratulations to all the people involved! Special mention to @aabadie @dylad and @MrKevinWeiss. Thanks a lot guys!

**Comment by kaspar030** [comment]
Yes. Congrats to getting the biggest (I think) API-rework in RIOT *ever* merged!

**Comment by dylad** [comment]
Wow, it's finally over ! I'm so relief !
I'll announce the new on the mailing list today.

**Comment by vincent-d** [comment]
Congrats and thanks to everyone!
